### ISSUE TITLE
[Feature Request] Consider additionally subclassing @beartype violations from `TypeError`

### ISSUE THREAD

**Issue by jdogburck** [beartype/beartype#155]

**Comment by leycec** [comment]
**Hmm...** They could, I suppose? I'm certainly not opposed. Doing so seems mostly harmless and maybe even beneficial. But is there any pragmatic real-world value-add to refactoring @beartype's type-checking violations to additionally subclass `TypeError`? Specifically, is there anything @beartype's current exception hierarchy *cannot* do that an exception hierarchy incorporating `TypeError` *can* do?

I'm mostly just curious. Please enlighten me, because I lack context and feel dumb. Bestow me with wisdom, ancient and erudite Pythonistas among us!

**Comment by jdogburck** [comment]
Technically there's nothing the current hierarchy can't do but it upended my expectations.   *Bear* with me and my, being relatively new to Python in my eyes, possible naivety...

It seems, the pragmatic real-world value in placing them under `TypeError` would be to place them under the standard exception to be 
>Raised when an operation or function is applied to an object of inappropriate type. ... Passing arguments of the wrong type (e.g. passing a [list](https://docs.python.org/3/library/stdtypes.html#list) when an [int](https://docs.python.org/3/library/functions.html#int) is expected) should result in a [TypeError](https://docs.python.org/3/library/exceptions.html?highlight=typeerror#TypeError), but passing arguments with the wrong value (e.g. a number outside expected boundaries) should result in a [ValueError](https://docs.python.org/3/library/exceptions.html?highlight=typeerror#ValueError).

The problem may come with that last clause, given the power of Beartype's Validators.   They can validate not only type (structure) as in `Annotated[str]` which seems to be in the purview of `TypeError` and it's descendents, but also value as in `Annotated[str, Is[lambda text: 4 <= len(text) <= 40]]` which strays into `ValueError` territory.

Full disclosure: I haven't reviewed Beartype's exception hierarchy in detail to see how the two cases are handled 😞

As I said, this may all be a misplaced surprise at not finding type errors somewhere under Python's `TypeError`.

let the elucidation continue...

**Comment by leycec** [comment]
**Enthralling points.** I *love* digging deep into typing semantics like this. The implementation possibilities, spirited debates, and productive discussions are endless!

Interestingly, the Python community has yet to formalize exception-handling semantics for runtime type checkers. There are *no* PEP standards – or even soft guidelines, really. It's still the Wild West out there. We kinda just made up our own exception hierarchy out of regurgitated cat hair and vellum.

> The problem may come with that last clause, given the power of Beartype's Validators.

Indeed, that is the rub.

> Full disclosure: I haven't reviewed Beartype's exception hierarchy in detail to see how the two cases are handled disappointed :disappointed: 

I myself no longer know what @beartype is doing, either. Let's see... just a hot minute... *shuffles around in the back of the Nissan Micra noisily*

And it looks like we're currently ambiguously conflating those two cases and just raising `BeartypeCallHintViolation` exceptions for both. That's not necessarily ideal, given what you've just suggested. We probably would not be able to productively separate `TypeError` from `ValueError` exceptions at the moment.

Let's go hypothetical. What if we *did* raise different types of exceptions for pure type violations and pure value violations? Could we reasonably distinguish these two cases for arbitrary `beartype.vale.Is[...]` functional validators? Functional validators are Turing-complete. Anyone can technically validate literally *anything* in a single `Annotated[{type}, Is[lambda obj: ...]]` type hint – including non-trivial combinations of multiple types and values: e.g.,

```python
from beartype import beartype
from beartype.typing import Annotated
from beartype.vale import Is
from collections.abc import Callable, Collection

# Type hint matching either:
# * Any callable defined in a REPL or script.
# * Any non-empty collection containing the string "__main__".
IsObjectMain = Annotated[object, Is[lambda obj:
    (isinstance(obj, Callable) and obj.__module__ == '__main__') or
    (isinstance(obj, Collection) and bool(obj) and '__main__' in obj)
]]

@beartype
def muh_crazy_func(obj: IsObjectMain) -> None: ...
```

What should @beartype raise if the `obj` parameter passed to `muh_crazy_func()` violates the `IsObjectMain` type hint? I... don't know. That type hint validates both types *and* values in a type-dependent manner.

Let's cogitate and possibly pretend this never happened. Thankfully, it's time to bicycle outside on the Great Canadian Trail! In the endless bliss of physical exercise, all our worries and our cares disappear like dopamine in the rain. :bicyclist:

**Comment by jdogburck** [comment]
ah - what a great way to start the day 🌞

with great power comes great responsibility *thwip* (is that webbing or a lasso?) ...    since we're just brainstorming - perhaps if @beartype differentiated combinators such that there were ones that were 

- *type-like* such as 'AnnotatedType' that only accepted types as parameters or things like `IsInstance[...]`, `IsSubClass[...]` a toned down `IsAttr[...]` (as it is it can wander outside of *type-like* things), etc. 
- *value-like* such as `AnnotatedValue` that accepted `IsEqual`, possibly comparers (e.g. `IsGt`, etc.  - urp - I think i just sicked up a bit) that don't exist yet 
- *muh_crazy-like* such as `Annotated` is today (where `Is[...]` lives and breathes)

Can we safely combine the 1st two using the type algebra in @beartype and stay in these, decidedly shaky, spaces?

This all makes me a bit woozy.   I'm enamored with the idea of a *type-like* (including ducks) `Annotation` that, if violated, raised TypeError.  It seems once you leave that space it's just everything else - as you say - the wild wild west.

Can we even define that space?  Have parts of the PEPs already wandered outside?

**Comment by leycec** [comment]
**Resolved by @beartype 0.17.0,** which offers a family of new `violation_type` configuration option. Behold the shocking truth!

```python
from beartype import BeartypeConf
from beartype.claw import beartype_this_package

# My spirit guide says that normalcy is a state of mind. Yet, you disagree.
beartype_this_package(conf=BeartypeConf(
    violation_type=AttributeError,     # <-- actually, never pass "AttributeError"
    violation_door_type=RuntimeError,  # <-- ..................or "RuntimeError"
    violation_param_type=TypeError,    # <-- Okay. Fine. This is okay.
    violation_return_type=ValueError,  # <-- passing "ValueError": not a great idea
))
```

See the [@beartype 0.17.0 announcement](https://github.com/beartype/beartype/discussions/317) for "documentation." Laziness prevails! :muscle: :bear:
